File: examples/javascript/src/renderer.js
/** Collects shapes, renders them, and provides filter/aggregate helpers. */
export class Renderer {
  #shapes = [];

  add(shape) {
    this.#shapes.push(shape);
    return this;
  }

  get count() { return this.#shapes.length; }

  render() {
    const header = `=== Renderer (${this.count} shapes) ===`;
    const body   = this.#shapes.map((s) => s.describe()).join("\n");
    const footer = `Total area: ${this.totalArea().toFixed(4)}`;
    return [header, body, footer].join("\n");
  }

  filter(predicate) {
    return this.#shapes.filter(predicate);
  }

  totalArea() {
    return this.#shapes.reduce((sum, s) => sum + s.area(), 0);
  }

  sortedByArea({ reverse = false } = {}) {
    const sorted = [...this.#shapes].sort((a, b) => a.area() - b.area());
    return reverse ? sorted.reverse() : sorted;
  }

  // Generator — yields shapes lazily
  *[Symbol.iterator]() {
    yield* this.#shapes;
  }
}

File: examples/python/src/renderer.py
"""Renderer: collects shapes, filters, and computes aggregates."""

from __future__ import annotations

from functools import reduce
from typing import Callable, Protocol

from shape import Shape


class HasArea(Protocol):
    """Structural typing — any object with an area() method."""

    def area(self) -> float: ...


FilterFn = Callable[[Shape], bool]


class Renderer:
    def __init__(self) -> None:
        self._shapes: list[Shape] = []

    def add(self, shape: Shape) -> None:
        self._shapes.append(shape)

    @property
    def count(self) -> int:
        return len(self._shapes)

    def render(self) -> str:
        header = f"=== Renderer ({self.count} shapes) ==="
        body = "\n".join(s.describe() for s in self._shapes)
        footer = f"Total area: {self.total_area():.4f}"
        return f"{header}\n{body}\n{footer}"

    def filter(self, predicate: FilterFn) -> list[Shape]:
        return [s for s in self._shapes if predicate(s)]

    def total_area(self) -> float:
        return reduce(lambda acc, s: acc + s.area(), self._shapes, 0.0)

    def sorted_by_area(self, *, reverse: bool = False) -> list[Shape]:
        return sorted(self._shapes, key=lambda s: s.area(), reverse=reverse)

    # Generator — yields shapes lazily
    def iter_shapes(self):
        yield from self._shapes
